httpcache.go

```diff
@@ -498,6 +513,49 @@ func headerAllCommaSepValues(headers http.Header, name string) []string {
 	return vals
 }
 
+// cachedReadCloser is a wrapper around a ReadCloser that calls onEOF
+// handler with a full copy of the content read from that ReadCloser
+// when EOF is read.
+type cachedReadCloser struct {
+	r io.ReadCloser
+	// buf stores a copy of the content of r.
+	buf bytes.Buffer
+	// onEOF is called with buf when EOF is read from r.
+	onEOF func(io.ReadCloser)
+}
+
+// newCachedReadCloser returns a new cachedReadCloser, that calls onEOF
+// with a copy of the content of r when EOF is read.
+func newCachedReadCloser(r io.ReadCloser, onEOF func(io.ReadCloser)) io.ReadCloser {
+	return &cachedReadCloser{
+		r:     r,
+		onEOF: onEOF,
+	}
+}
+
+// Read reads the next len(p) bytes from r or until r is drained. The
+// return value n is the number of bytes read. If r has no data to
+// return, err is io.EOF (unless len(p) is zero) and onEOF is called
+// with a full copy of what has been read so far.
+func (r *cachedReadCloser) Read(p []byte) (n int, err error) {
+	n, err = r.r.Read(p)
+	if n > 0 {
+		if n, err := r.buf.Write(p[:n]); err != nil {
+			return n, err
+		}
+	}
+	if err == io.EOF {
+		buf := r.buf
+		r.onEOF(ioutil.NopCloser(&buf))
+	}
+
+	return
```

Comment:
I think naked returns are bad for readability, I would suggest being explicit about what's being returned:

```Go
return n, err
```